fedimint-core/src/db/mod.rs

```diff
@@ -1841,28 +1861,127 @@ pub async fn apply_migrations(
         while current_db_version < target_db_version {
             if let Some(migration) = migrations.get(&current_db_version) {
                 info!(target: LOG_DB, "Migrating module {kind} from {current_db_version} to {target_db_version}");
-                migration(&mut dbtx.to_ref_nc()).await?;
+                if let Some(module_instance_id) = module_instance_id {
+                    migration(
+                        &mut global_dbtx
+                            .to_ref_with_prefix_module_id(module_instance_id)
+                            .into_nc(),
+                    )
+                    .await?;
+                } else {
+                    migration(&mut global_dbtx.to_ref_nc()).await?;
+                }
             } else {
                 panic!("Missing migration for version {current_db_version}");
             }
 
             current_db_version.increment();
-            dbtx.insert_entry(&DatabaseVersionKey, &current_db_version)
+            global_dbtx
+                .insert_entry(
+                    &DatabaseVersionKey(module_instance_id_key),
+                    &current_db_version,
+                )
                 .await;
         }
 
         current_db_version
     } else {
-        dbtx.insert_entry(&DatabaseVersionKey, &target_db_version)
-            .await;
         target_db_version
     };
 
-    dbtx.commit_tx_result().await?;
+    global_dbtx.commit_tx_result().await?;
     info!(target: LOG_DB, "{} module db version: {}", kind, db_version);
     Ok(())
 }
 
+/// Migrates the `DatabaseVersion` from inside the module's isolated database
+/// to the global namespace. Database migrations are driven from outside of the
+/// modules so the `DatabaseVersion` should also exist outside the modules.
+/// This also fixes the key prefix conflicts with 0x50.
+pub async fn migrate_database_version(
+    db: &Database,
+    target_db_version: DatabaseVersion,
+    module_instance_id: Option<ModuleInstanceId>,
+    kind: String,
+) -> Result<(), anyhow::Error> {
+    db.ensure_global()?;
+    let mut global_dbtx = db.begin_transaction().await;
+
+    // Use 0xff for fedimint-server and the `module_instance_id` for each module
+    let key_module_instance_id = if let Some(module_instance_id) = module_instance_id {
+        module_instance_id
+    } else {
+        MODULE_GLOBAL_PREFIX.into()
+    };
+
+    // First check if the module has a `DatabaseVersion` written to
+    // `DatabaseVersionKey` If `DatabaseVersion` already exists, there is
+    // nothing to do.
+    if global_dbtx
+        .get_value(&DatabaseVersionKey(key_module_instance_id))
+        .await
+        .is_none()
+    {
+        // Read and remove the previous `DatabaseVersion`, which is in the module's
+        // isolated namespace (but not for fedimint-server)
+        let current_version_in_module = if let Some(module_instance_id) = module_instance_id {
```

Comment:
Sanity check: verify that the value has a length we expect, panic otherwise?

Reply:
`DatabaseVersion` is just a `u64`. Are we doing variable length encoding of integers? Is it possible to this to ever not be 8 bytes?

Reply:
I'm thinking: is it possible someone could have a key `0x50` and didn't ever do a migration, so that migration code would mistake their value for the version? Unlikely, but is it impossible?

> Are we doing variable length encoding of integers? 

Yes. Thought I don't think anyone will have > 253 or so migration versions to start getting into 3 byte encodings. :)

Reply:
Hmm yes I suppose this is possible. A module would need to write a key/value using the 0x50 prefix and then execute this migration.

Although I believe in that case the decoding of the `DatabaseVersion` should fail (unless it also coincidentally happens to be a `u64`, in which case I don't think there's much we can do) and there is already code that asserts on decode failure. So I think this length check is already implicitly covered.

Reply:
> Although I believe in that case the decoding of the `DatabaseVersion` should fail (unless it also coincidentally happens to be a `u64`, in which case I don't think there's much we can do)

You're right. In that case it's good enough.